Extended documentation for upgrading from CI3 to CI4 (#4338)
codeigniter4/CodeIgniter4: user_guide_src/source/installation/upgrade_migrations.rst

Upgrade Migrations

> ###############################################################################

Reviewer: Adjust the lines in this file accordingly.

.. contents::
    :local:
    :depth: 1

Documentations
============================================================

- `Database Migrations Documentation Codeigniter 3.X <http://codeigniter.com/userguide3/libraries/migration.html?highlight=migration>`_

> - `Database Migrations Documentation Codeigniter 4.X <http://codeigniter.com/user_guide/dbmgmt/migration.html?highlight=migration>`_

Reviewer 2: ```suggestion
- `Database Migrations Documentation Codeigniter 4.X </dbmgmt/migration.html?highlight=migration>`_
```

What has been changed
============================================================

> - First of all the sequential naming (``001_create_users``, ``002_create_posts``) of migrations is not longer supported. Version 4 of CodeIgniter only supports the timestamp scheme (``20121031100537_create_users``, ``20121031500638_create_posts``) . If you have used sequential naming you have to rename each migration file.
> - The migration procedure has been also changed. You can now migrate the database with a simple cli command::
> 
>     > php spark migrate

Reviewer: ```suggestion
- First of all, the sequential naming (``001_create_users``, ``002_create_posts``) of migrations is not longer supported. Version 4 of CodeIgniter only supports the timestamp scheme (``20121031100537_create_users``, ``20121031500638_create_posts``) . If you have used sequential naming you have to rename each migration file.
- The migration procedure has been also changed. You can now migrate the database with a simple CLI command::

    > php spark migrate
```

Upgrade Guide
============================================================
1. If your v3 project uses sequential migration names you have to change those to timestamp names.

> 2. You have to move all migration files to the new folder ``app/Database/Migrations``
> 3. Remove the following line ``defined('BASEPATH') OR exit('No direct script access allowed');``
> 4. Add this line just after the opening php tag: ``namespace App\Database\Migrations;``
> 5. Below the ``namespace App\Database\Migrations;`` line add this line: ``use CodeIgniter\Database\Migration;``
> 6. Replace ``extends CI_Migration`` with ``extends Migration``
> 7. The method names within the dbforge class has been changed. For example:

Reviewer: ```suggestion
2. You have to move all migration files to the new folder ``app/Database/Migrations``.
3. Remove the following line ``defined('BASEPATH') OR exit('No direct script access allowed');``.
4. Add this line just after the opening php tag: ``namespace App\Database\Migrations;``.
5. Below the ``namespace App\Database\Migrations;`` line add this line: ``use CodeIgniter\Database\Migration;``
6. Replace ``extends CI_Migration`` with ``extends Migration``.
7. The method names within the ``Forge`` class has been changed to use camelCase. For example:
```

- ``$this->dbforge->add_field`` to ``$this->forge->addField``
- ``$this->dbforge->add_key`` to ``$this->forge->addKey``
- ``$this->dbforge->create_table`` to ``$this->forge->addTable``
- ``$this->dbforge->drop_table`` to ``$this->forge->addTable``

8. (optional) You can change the array syntax from ``array(...)`` to ``[...]``

Code Example
============================================================

Codeigniter Version 3.11
-------------------------------------------------------
Path: ``application/migrations``::

    <?php

    defined('BASEPATH') OR exit('No direct script access allowed');

>     class Migration_Add_blog extends CI_Migration {
> 

Reviewer: ```suggestion
    class Migration_Add_blog extends CI_Migration
    {
```

                                    'type' => 'INT',
                                    'constraint' => 5,
                                    'unsigned' => TRUE,
                                    'auto_increment' => TRUE
                            ),
                            'blog_title' => array(
                                    'type' => 'VARCHAR',
                                    'constraint' => '100',
                            ),
                            'blog_description' => array(
                                    'type' => 'TEXT',
                                    'null' => TRUE,
                            ),
                    ));
                    $this->dbforge->add_key('blog_id', TRUE);
                    $this->dbforge->create_table('blog');
            }

            public function down()
            {
                    $this->dbforge->drop_table('blog');
            }
    }

Codeigniter Version 4.x
-------------------------------------------------------
Path: ``app/Database/Migrations``::

    <?php
    namespace App\Database\Migrations;

    use CodeIgniter\Database\Migration;

    class AddBlog extends Migration
    {

> 

Reviewer: ```suggestion
```
